### PR TITLE
Homework - Cars

### PR DESCRIPTION
**Pull request by dawidjozefczak**
- Exceptions for InvalidGear, InvalidSpeed and InvalidEngine.
- Encapsulation class member are private.
- Check car velocity is zero to change engine.
- Memory leaks eliminated completely by using smart pointers.
- Tests added.
- Workflow file added

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Good job 👍

File: PetrolEngine.cpp [hunk 1]
```diff
@@ -1,20 +1,38 @@
...
 
-void PetrolEngine::changeGear(int gear)
-{
+void PetrolEngine::changeGear(int gear) {
+    if (gear > gears_ || gear < -1) {
+        throw InvalidGear("Gearbox out of range!");
+    }
+    if (gear == -1 && currentGear_ >= 2) {
+        throw InvalidGear("Gearbox cannot change gear to R from such a high gear");
+    }
```
Comment (WBigaj) [inline PetrolEngine.cpp:16]: You should make gears const values instead of magic ones.

File: PetrolEngine.cpp [hunk 2]
```diff
@@ -1,20 +1,38 @@
...
     // TODO: Add checking if gear is between -1 and gears_
     // -1 is for REAR
     // 0 is for NEUTRAL
```
Comment (WBigaj) [inline PetrolEngine.cpp:19]: I think this comment can be removed.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
4 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**